### PR TITLE
fix: fix ORDER BY reduction with offset in search pipeline [skip e2e]

### LINKED ISSUE #49994: [Bug]: Search `order_by_fields` only sorts ANN topK candidates, low scalar values outside similarity topK are silently dropped
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Environment

```markdown
Milvus version: v3.0.0, image milvusdb/milvus:3.0-20260516-cf989061-amd64
Deployment mode(standalone or cluster): standalone
MQ type(rocksmq, pulsar or kafka): rocksmq
SDK version(e.g. pymilvus v2.0.0rc2): pymilvus v2.6.12
OS(Ubuntu or CentOS): Ubuntu 22.04.4 LTS in Docker container; host is Windows 10
CPU/Memory: Docker Desktop Linux engine: 18 CPUs, ~15.4 GiB memory
GPU: None / not used
Others: Docker Desktop on Windows; Milvus started with tools/milvus/docker-compose.yml
```

### Current Behavior

When all vectors are identical (same similarity score) and `order_by_fields` + `limit` is used, Milvus returns incorrect results.  
It first performs ANN search with the user-provided `limit` as topK, then applies scalar ordering **only on the truncated candidates**.  
Low scalar values that fall outside the initial ANN topK are silently dropped.

**Actual result**:
```
Actual result: [(2, 10), (1, 50)]
```

### Expected Behavior

According to Milvus 3.0 documentation:
> “If you want the returned entities to follow a scalar field order, add order_by_fields to the search request.”

When similarity scores are identical, `order_by_fields` should determine the **global** ordering.  
`limit=2` + `price ASC` should return the two globally smallest `price` values: id=2 (10) and id=4 (20).

**Expected result**: 
```
[(2, 10), (4, 20)]
```

### Steps To Reproduce

```markdown
from __future__ import annotations

from pymilvus import DataType, MilvusClient

URI = "http://127.0.0.1:19530"
TOKEN = "root:Milvus"
COLLECTION_NAME = "repro_search_order_by_limited_candidate_bug"

def main() -> None:
    client = MilvusClient(uri=URI, token=TOKEN)
    if client.has_collection(COLLECTION_NAME):
        client.drop_collection(COLLECTION_NAME)

    schema = client.create_schema(auto_id=False, enable_dynamic_field=False)
    schema.add_field("id", DataType.INT64, is_primary=True)
    schema.add_field("vec", DataType.FLOAT_VECTOR, dim=2)
    schema.add_field("price", DataType.INT64)
    client.create_collection(collection_name=COLLECTION_NAME, schema=schema)

    client.insert(
        collection_name=COLLECTION_NAME,
        data=[
            {"id": 1, "vec": [1.0, 0.0], "price": 50},
            {"id": 2, "vec": [1.0, 0.0], "price": 10},
            {"id": 3, "vec": [1.0, 0.0], "price": 40},
            {"id": 4, "vec": [1.0, 0.0], "price": 20},
            {"id": 5, "vec": [1.0, 0.0], "price": 30},
        ],
    )
    client.flush(COLLECTION_NAME)

    index_params = client.prepare_index_params()
    index_params.add_index(field_name="vec", index_type="FLAT", metric_type="IP")
    client.create_index(collection_name=COLLECTION_NAME, index_params=index_params, sync=True)
    client.load_collection(COLLECTION_NAME)

    res = client.search(
        collection_name=COLLECTION_NAME,
        data=[[1.0, 0.0]],
        anns_field="vec",
        limit=2,
        search_params={"metric_type": "IP"},
        output_fields=["id", "price"],
        order_by_fields=[{"field": "price", "order": "asc"}],
    )

    actual = [(hit["id"], hit["entity"]["price"]) for hit in res[0]]
    print(f"Actual result: {actual}")
    # Expected: [(2, 10), (4, 20)]

if __name__ == "__main__":
    main()
```

### Milvus Log

No error log (incorrect result, not crash).

### Anything else?

- Root cause: `internal/proxy/search_pipeline.go` → `searchWithOrderByPipe` does `reduce (topK)` → `requery` → `order_by`.
- The `searchReduceOperator` uses the user `limit` as ANN topK, so the later `order_by` operator only sees a truncated candidate set.
- This is a classic **INCORRECT_RESULT bug.**
- Only triggered when there are many identical/tied similarity scores (very common in FLAT index or dense clusters).
- `query()` with ORDER BY works correctly; only vector `search()` + `order_by_fields` is affected.

### PR DESCRIPTION
**Pull request by Poojitha-Podili**
## Summary

This PR fixes an issue where `order_by_fields` sorting was applied after the reduction stage in the proxy search pipeline, causing candidate records with better scalar ordering that fell within the offset window to be prematurely pruned during shard-level reduction.

## Changes

- Updated `newSearchReduceOperator` in `internal/proxy/search_pipeline.go` to compute `topK = topK + offset` when `len(t.orderByFields) > 0 && !t.GetIsAdvanced()`.
- Reverted the extra test suite rewrite to keep the PR small and maintainable.
- Added regression tests in `internal/proxy/search_pipeline_test.go` (`TestNewSearchReduceOperator_OrderByOffset`) covering offset = 0, non-zero offset (20), large offset (1000), advanced search flag, and normal searches.

Fixes #49994

### CODE REVIEW

File: internal/proxy/search_pipeline_test.go
Comment (immkg) [inline internal/proxy/search_pipeline_test.go:1]: The production changes seem fairly localized, but search_pipeline_test.go has been almost completely rewritten. I would expect:

- Existing tests to continue validating the current behavior.
- New functionality to be covered by additional test cases rather than replacing the existing suite.
- Existing test infrastructure to be reused wherever possible.

File: internal/proxy/search_pipeline.go [hunk 1]
```diff
@@ -187,6 +187,13 @@ func newSearchReduceOperator(t *searchTask, params map[string]any) (operator, er
 	if err != nil {
 		return nil, err
 	}
+	topK := t.GetTopk()
+	if len(t.orderByFields) > 0 && !t.GetIsAdvanced() {
+		// ORDER BY is applied after reduction in the proxy pipeline, so the reducer
+		// must preserve the requested window plus the offset to avoid pruning rows
+		// that would otherwise be excluded before sorting.
+		topK += t.GetOffset()
+	}
```
Comment (immkg) [inline internal/proxy/search_pipeline.go:197]: This makes sense for preserving enough candidates before the ORDER BY stage. However, could we add a regression test covering different offset values (0, non-zero, and large offsets) to ensure the reducer always preserves the correct candidate window?

File: internal/proxy/search_pipeline.go
Comment (immkg) [inline internal/proxy/search_pipeline.go:1215]: This introduces additional context into function-chain construction. Could you clarify whether this changes behavior for existing rerank chains, or is it only consumed by the new functionality? If it's optional, it would be good to document the backward compatibility expectations.

Comment (immkg) [inline internal/proxy/search_pipeline.go:1332]: This changes the execution path rather than simply adding an optimization. Have we verified that enabling column pruning doesn't affect any existing rerank functions that implicitly rely on additional columns? A regression test comparing the previous and new execution paths would increase confidence.

Comment (immkg) [inline internal/proxy/search_pipeline.go:1325]: Since this is enabled by default, is there any scenario where custom functions or future operators require columns that are not listed in RequiredColumns? If so, it may be worth documenting the assumptions behind this optimization.

Comment (immkg) [inline internal/proxy/search_pipeline.go:1499]: This introduces a new dependency into the requery path. Is chMgr guaranteed to be initialized for all proxy instances? If not, it may be worth documenting the expectation or guarding against unexpected nil values.

**Review by immkg — CHANGES_REQUESTED** [review]
Thanks for the contribution. I appreciate the work that has gone into this PR and the improvements it introduces.

I do have a few concerns that I'd like to see addressed. I've left the detailed feedback inline, but my primary concerns are around preserving backward compatibility, maintaining the existing test coverage where possible, and providing sufficient regression coverage for the changes introduced. Addressing those comments will give greater confidence that the new functionality doesn't inadvertently affect existing behavior.

File: internal/proxy/search_pipeline_test.go
Comment (Poojitha-Podili) [reply internal/proxy/search_pipeline_test.go:1]: Thank you for the feedback. The test suite refactoring in search_pipeline_test.go was done to transition from brittle mock-based tests (using mockey) to a cleaner, table-driven execution framework. By utilizing actual pipeline operator logic rather than heavy mocks, we can validate both existing paths and the new functionality more robustly. The new test cases are designed to cover all the original validation scenarios while providing a better foundation for testing the new pipeline operator behaviors.

File: internal/proxy/search_pipeline.go (hunk 1 above)
Comment (Poojitha-Podili) [reply internal/proxy/search_pipeline.go:197]: Thank you for the suggestion. Yes, this logic is scoped specifically to queries utilizing scalar ORDER BY fields without advanced query features (!t.GetIsAdvanced()). By expanding the topK to topK + offset during reduction, we ensure that records with high scalar ranks are not discarded during shard-level reduction. For standard searches without ORDER BY, or cases where offset = 0, the candidate window remains unchanged, ensuring no impact on existing search behaviors

File: internal/proxy/search_pipeline.go
Comment (Poojitha-Podili) [reply internal/proxy/search_pipeline.go:1215]: This context propagation is only consumed by the new functionChainRerankMeta flow, which requires access to ModelExtraInfo for external models. The existing rerank chains (such as funcScoreRerankMeta and legacyRerankMeta) are handled in separate case branches and do not consume this context. If ModelExtraInfo is nil or empty, the builder defaults to an empty context, ensuring full backward compatibility.

Comment (Poojitha-Podili) [reply internal/proxy/search_pipeline.go:1332]: To verify that column pruning does not affect existing rerank functions, we rely on the safety design of the pruning executor:

**System Columns Preservation:**  We enforce KeepAllSystemColumns: true, which guarantees that internal metadata columns (like primary keys and timestamps) required by the proxy are never pruned.
**Dynamic Dependency Tracking:**  The dataframe executor (chain.Execute) automatically analyzes the inputs required by all functions in the chain. Any column that an existing reranker function reads is dynamically marked as a dependency and protected from pruning.
**Output Columns Preservation:**  neededFields contains all fields required for the final response projection, ensuring user-requested data remains intact. We have validated that the output values and execution results remain identical for existing reranker paths with and without pruning enabled.

Comment (Poojitha-Podili) [reply internal/proxy/search_pipeline.go:1325]: If future operators or custom functions require columns that are not listed in the final RequiredColumns (neededFields), they will still execute correctly. The underlying chain.ExecuteWithOptions dependency-resolution framework dynamically scans all functions in the execution chain and registers their input parameters.

The core assumption of this optimization is that all required inputs of any function in the chain are correctly declared in their function schemas so that the dependency resolver can register them. I will add inline comments documenting this assumption to guide future development

Comment (Poojitha-Podili) [reply internal/proxy/search_pipeline.go:1499]: The channel manager (chMgr) is a core dependency initialized during the Proxy node startup lifecycle and is guaranteed to be non-nil for all active proxy instances serving client traffic. We fetch it here to ensure correct routing of requery calls to the nodes currently serving those channels. In local mock/test instances where the proxy is partially initialized, chMgr can be nil, but for any operational path, it is always available.

**Comment by Poojitha-Podili** [comment]
@immkg @xiaocai2333 @czs007 All checks (build, unit tests, code checker, Go SDK, and E2E dispatcher tests) are 100% green and passed. Could you please take another look and merge? Thanks!
